### ISSUE TITLE
Convolution<> Layer cauing Segmentation Fault( core dumped) error.

### ISSUE THREAD

**Issue by prince776** [mlpack/mlpack#2310]
Hi, I was working on developing models and whenever I exceeded a certain limit of output maps in convolution layer, and then tried to train, I got the error : `Segmentation fault (core dumped)`
basically:
```
model.Add<Convolution<>>(3, 16, 7, 7, 2, 2, 3, 3, 28,28);
model.Add<Convolution<>>(16, 192, 3, 3, 1, 1, 1, 1, 10,10);
...
model.Train(Xs,Ys,optimizer);
```
works, but
```
model.Add<Convolution<>>(3, 64, 7, 7, 2, 2, 3, 3, 28,28);
model.Add<Convolution<>>(64 192, 3, 3, 1, 1, 1, 1, 10,10);
...
model.Train(Xs,Ys,optimizer);
```
terminates with error: 
```
Segmentation fault (core dumped)
```

As suggested by @zoq, I used gdb to find what exactly is causing error and I found that
```
arma::Mat<double>::operator=(arma::Mat<double> const&) ()
```
inside  `Gradient()` function of `Convolution<>` class is causing this error when called during Training process, which probably means when a matrix is being filled, but where exactly I have no idea.

Full backtrace: https://pastebin.com/bvVzJJjs

I'm not able to figure out further what to try.
Any help/suggestion are welcome.

**Comment by codeboy5** [comment]
I am facing similar issues with the ZFNet model I have implemented.

**Comment by kartikdutt18** [comment]
@prince776, Could you try one more thing please. Could run predict without training, if the problem is associated with Gradient then at least predict part should work fine. Thanks a lot for opening this issue and I think the above simple check might give us more insight. Thanks a ton.

**Comment by prince776** [comment]
Hi @codeboy5  @kartikdutt18 Thanks for the responses. Yes @kartikdutt18 Predict is working totally fine. Just as gbd showed, it's due to Gradient Function of Convolution class.

**Comment by kartikdutt18** [comment]
Okay, I'll try to debug it tonight / early tomorrow. Whatever find I'll post it here. Thanks.

**Comment by prince776** [comment]
How did you run AlexNet by the way? I saw your implementation of AlexNet it also had 64 output maps. You were using Sequential<> instead of FFN<>, can you tell me how did you train it?

**Comment by himanshupathak21061998** [comment]
@prince776 Thanks for posting this @kartikdutt18 I will also try to debug this if I find something I will post it here.

**Comment by kartikdutt18** [comment]
@prince776, I think as far as I remember for AlexNet I used padding and FFN type implementation rather than a class type. The class type always failed so I assumed that I had misused a pointer somewhere. Also one thing to know is **You need to add Identity layer before sequential if sequential is your first layer.** Thanks a lot @himanshupathak21061998, I think this way if there is really an issue we can get it resolved as quickly as possible.

**Comment by kartikdutt18** [comment]
So far this is what I have done:
1. Built a neural network test case to add in mlpack.
2. Tested it for various output channels. (2, 8, 16, 32).
3. Result : Worked for 2, 8, 16. 
Test is stuck on 32 channel output.
Also there are model implementation of VGG19, 16 which have channels upto 512. And if they do work then it maybe it is related to limitation of our device.

**Comment by kartikdutt18** [comment]
Okay so the test didn't segfault on 32 channels. Going for vgdb.

**Comment by prince776** [comment]
I was also thinking if it's due to some sort of device limitation, but while running the program the memory and CPU usage all were not nearly full. 
I think you said you have a raspberry pi set up with mlpack so, can you try to see at how many output channels does it segfaults compared with your laptop/computer.

**Comment by kartikdutt18** [comment]
I can try that but before I go for it, I think @adithya-tp has implemented VGG16, so I wanted to know if he could share some results to ensure that it works. That way we know that it's not related to the device. Could you also share details of your system:
My Local System has the following config:
1. 16 GB RAM
2. i7, 7 th gen (2.3 GHz)
3. OS: OSX (Catalina)

**Comment by prince776** [comment]
My Local system has the following specifications:
1. 8 GB RAM
2. i5 8th gen (2.3 GHz)
3. OS: Ubuntu 16.04
4. Dedicated Nvidia GTX 1050 Ti, 4GB VRAM (I wasn't using BLAS though)

**Comment by rcurtin** [comment]
If you apply the changes from https://github.com/mlpack/mlpack/pull/2234, does it fix the issue?

**Comment by kartikdutt18** [comment]
I am building mlpack using that right now, Hopefully will get to know about that soon.

**Comment by kartikdutt18** [comment]
Hey @prince776, I took my LeNet and made the following change:

```
  leNet = new Sequential<>();
  ConvolutionBlock(inputChannel, 6, 5, 5, 1, 1, 2, 2);
  PoolingBlock(2, 2, 2, 2);
  ConvolutionBlock(6, 512, 5, 5, 1, 1, 2, 2);
  PoolingBlock(2, 2, 2, 2);
  // Add linear layer for LeNet.
  if (leNetVer == 1)
  {
    leNet->Add<Linear<>>(512 * inputWidth * inputHeight, numClasses);
  }
```
It's working fine for me (almost 1 epoch complete, will try higher channels and get back to you).
However alexnet fails so maybe the problem isn't with convolution?
Hopefully I will know more in an hour.

**Comment by kartikdutt18** [comment]
Other thing I tried, I reduced kernel sizes in alexnet and printed input shape after each convolution.
```
16
8
4
4
5
```
But it segfaults even after creating a copy for inputs in convolution layer.

**Comment by prince776** [comment]
> Hey @prince776, I took my LeNet and made the following change:
> 
> ```
>   leNet = new Sequential<>();
>   ConvolutionBlock(inputChannel, 6, 5, 5, 1, 1, 2, 2);
>   PoolingBlock(2, 2, 2, 2);
>   ConvolutionBlock(6, 512, 5, 5, 1, 1, 2, 2);
>   PoolingBlock(2, 2, 2, 2);
>   // Add linear layer for LeNet.
>   if (leNetVer == 1)
>   {
>     leNet->Add<Linear<>>(512 * inputWidth * inputHeight, numClasses);
>   }
> ```
> 
> It even works with 4048.
> It's working fine for me (almost 1 epoch complete, will try higher channels and get back to you).
> However alexnet fails so maybe the problem isn't with convolution?
> Hopefully I will know more in an hour.

Can you share the full code of this please.

**Comment by kartikdutt18** [comment]
The code is available [here](https://github.com/mlpack/models/pull/60).Changes that you need to make are:
1. There is an incomplete if condition in data loader, remove that.
2. replace output channel = 16 there by any number you want and make the change in `if` condition to match that number.
3. Run using `./bin/mnist_tutorial_cnn`

**Comment by prince776** [comment]
Thanks

**Comment by prince776** [comment]
Quick update:
I used Sequential<> to add convolution layers then wraped it in FFN with identity layer first (just like in your example of mnist) and it worked.
However if convolution layer was directly added to FFN and not to Sequential<> then it terminates with segfault.

So:
```
  Sequential<> module;
  module.Add<Convolution<>>(1, 64, 5, 5, 1, 1, 0, 0, 28,28);
  module.Add<LeakyReLU<>>();
  module.Add<MaxPooling<>>(2, 2, 2, 2);
  module.Add<Convolution<>>(64, 128, 5, 5, 1, 1, 0, 0, 12,12);

  FFN<NegativeLogLikelihood<>, RandomInitialization> model;
  model.Add<IdentityLayer<>>();
  model.Add<Sequential<>>(module);
  model.Train(Xs,Ys);
```
works.
but, 
```
  FFN<NegativeLogLikelihood<>, RandomInitialization> model;
  model.Add<Convolution<>>(1, 64, 5, 5, 1, 1, 0, 0, 28,28);
  model.Add<LeakyReLU<>>();
  model.Add<MaxPooling<>>(2, 2, 2, 2);
  model.Add<Convolution<>>(64, 128, 5, 5, 1, 1, 0, 0, 12,12);
  model.Train(Xs,Ys);
```
gives segfault error

**Comment by kartikdutt18** [comment]
Hey @prince776, Did you try out my LeNet yet?
I am really happy with that model, it worked everywhere so if you would like you can try that too. What do you think could be causing this, maybe we missed some documentation?
Like Identity layer bit that I said was written as a comment in sequential.hpp .

**Comment by prince776** [comment]
Update:
FFN also works without Sequential if IndentityLayer<> is added before everything. However it is slower.

**Comment by prince776** [comment]
> Hey @prince776, Did you try out my LeNet yet?
> I am really happy with that model, it worked everywhere so if you would like you can try that too. What do you think could be causing this, maybe we missed some documentation?
> Like Identity layer bit that I said was written as a comment in sequential.hpp .

I am working on it. Getting error: `[FATAL] Cannot open file './../data/mnist_train.csv'`. Problem is it still persists even if I add mnist_train.csv file

**Comment by kartikdutt18** [comment]
Run this, `cmake -DDOWNLOAD_MNIST=ON`

**Comment by prince776** [comment]
It works correctly.

**Comment by kartikdutt18** [comment]
Other option is `-DDOWNLOAD_DATASETS=ON` (It will download all datasets though). I am working on a dataloader to download file during runtime rather than when it is compiled.

**Comment by kartikdutt18** [comment]
Great!!!

**Comment by prince776** [comment]
So Apparently as long as IndentityLayer<> is added before these convolution layers, it works fine, but without it can terminate with segfault for some 50+ outputChannels.

**Comment by prince776** [comment]
> If you apply the changes from #2234, does it fix the issue?

This fixed the issue for me.

**Comment by kartikdutt18** [comment]
@prince776, Should we close this?

**Comment by himanshupathak21061998** [comment]
Oh, I am late @kartikdutt18 and @prince776  thanks for working on this. Maybe next time I will be able to help you guys.

**Comment by prince776** [comment]
> @prince776, Should we close this? If this issue is resolved for you.

I will close it later tomorrow because I'm still getting some segfaults in inception network. I'm not sure if that's because of my implementation or convolution layer. I will be doing further testing and once I'm done I'll close it if it's resolved.

**Comment by prince776** [comment]
I'm closing this issue, #2234 solves it.
